File: src/ssh_manager/providers/ssh_generic.py
"""Generic SSH adapter - ssh-copy-id / authorized_keys (any reachable server).

The universal fallback (resolution order #2): works on any server you
can already reach. ``deploy`` uses ``ssh-copy-id`` (idempotent); ``remove`` does
the whole read-modify-write in ONE remote invocation under an ``flock`` advisory
lock - it backs up ``authorized_keys`` first, removes the key by base64 body,
refuses to leave the file with no key lines (lockout guard), and writes atomically
(temp + ``mv``). Doing it in a single locked remote script avoids the race two
separate read-then-write ssh calls would have. Hardening adapted from a standalone
VPS key tool.
"""
from __future__ import annotations

from ..core.authorized_keys import key_body, key_lines
from ..util import proc
from .base import DeployOutcome, Provider, Target

_HINT = "install OpenSSH (macOS ships it; else: brew install openssh)"

# Remote read-modify-write of authorized_keys, run as one ssh command under flock.
# Exit codes: 0 removed; 2 key not present; 3 lockout guard tripped (would leave no
# key lines); 4 lock/setup failure. {body} is a base64 body (single-quoted, escaped).
_REMOVE_SCRIPT = r"""set -eu
AK="$HOME/.ssh/authorized_keys"
mkdir -p "$HOME/.ssh"; chmod 700 "$HOME/.ssh"
[ -f "$AK" ] || : > "$AK"
exec 9>"$HOME/.ssh/.ssh-manager.lock" || exit 4
if command -v flock >/dev/null 2>&1; then flock 9 || exit 4; fi
BODY='{body}'
grep -qF -- "$BODY" "$AK" || exit 2
TMP="$(mktemp "$AK.ssh-manager.XXXXXX")" || exit 4
grep -vF -- "$BODY" "$AK" > "$TMP" || true
# Lockout guard: refuse to leave authorized_keys with no real (non-comment) line.
if ! grep -Eq '^[[:space:]]*[^[:space:]#]' "$TMP"; then rm -f "$TMP"; exit 3; fi
cp -p "$AK" "$AK.ssh-manager.bak.$(date +%Y%m%d-%H%M%S)" 2>/dev/null || true
chmod 600 "$TMP"
mv "$TMP" "$AK"
"""


class GenericSSH(Provider):
    name = "generic-ssh"
    category = "server"

    @staticmethod
    def _known_hosts_opts(target: Target) -> list[str]:
        # Use this host's per-profile trust store, so deploy/verify read+populate the
        # SAME known_hosts that `ssh <alias>` uses (per-profile isolation) - not the
        # default ~/.ssh/known_hosts. accept-new (set by callers) then lands a
        # first-seen host key in the right store.
        return ["-o", f"UserKnownHostsFile={target.known_hosts}"] if target.known_hosts else []

    def _ssh_base(self, target: Target) -> list[str]:
        # ConnectTimeout so an unreachable host fails fast instead of hanging.
        base = ["ssh", "-o", "StrictHostKeyChecking=accept-new",
                "-o", "ConnectTimeout=10", *self._known_hosts_opts(target)]
        if target.port != 22:
            base += ["-p", str(target.port)]
        return base

    def deploy(self, target: Target) -> DeployOutcome:
        proc.require("ssh-copy-id", _HINT)
        # -o ConnectTimeout so an unreachable host fails fast instead of hanging
        # on TCP connect (ssh-copy-id forwards -o options to ssh).
        cmd = ["ssh-copy-id", "-o", "ConnectTimeout=10", *self._known_hosts_opts(target),
               "-i", str(target.pubkey_path)]
        if target.port != 22:
            cmd += ["-p", str(target.port)]
        cmd.append(target.ssh_dest)
        # Interactive: ssh-copy-id may prompt for a password (no existing key yet).
        # Bounded so a wedged connection (e.g. VPN-gated host on :443) can't hang.
        code = proc.run_interactive(cmd, timeout=120)
        if code != 0:
            return DeployOutcome(method="ssh-copy-id", verified=False,
                                 detail=f"ssh-copy-id exited {code}", error=True)
        return DeployOutcome(method="ssh-copy-id", verified=True,
                             detail="authorized_keys updated")

    def verify(self, target: Target) -> bool:
        if target.identity_path is None:
            return False
        cmd = [
            "ssh", "-i", str(target.identity_path),
            "-o", "BatchMode=yes", "-o", "IdentitiesOnly=yes",
            "-o", "StrictHostKeyChecking=accept-new", *self._known_hosts_opts(target),
        ]
        if target.port != 22:
            cmd += ["-p", str(target.port)]
        cmd += [target.ssh_dest, "true"]
        return proc.run(cmd, timeout=20).returncode == 0

    def list_deployed(self, target: Target) -> list[str]:
        return key_lines(self._read_authorized_keys(target) or "")

    def remove(self, target: Target) -> bool:
        """Revoke the key (best-effort, with safety rails). Returns True only if it
        actually removed the key. The whole read-modify-write happens in one remote
        script under flock, so it can't race another writer; rc!=0 (key absent,
        lockout guard, lock failure, or host unreachable) means it did not act."""
        body = key_body(target.pubkey_text)
        if not body:
            return False
        safe_body = body.replace("'", "'\\''")
        script = _REMOVE_SCRIPT.replace("{body}", safe_body)
        cmd = [*self._ssh_base(target), target.ssh_dest, script]
        return proc.run(cmd, timeout=30).returncode == 0

    # transport
    def _read_authorized_keys(self, target: Target) -> str | None:
        # The remote `|| true` makes a missing file a clean empty read (rc 0); a
        # non-zero rc therefore means ssh itself failed (unreachable/auth) - None,
        # distinct from a genuinely empty authorized_keys ("").
        cmd = [*self._ssh_base(target), target.ssh_dest,
               "cat ~/.ssh/authorized_keys 2>/dev/null || true"]
        r = proc.run(cmd, timeout=20)
        return r.stdout if r.returncode == 0 else None


# re-export for callers that match keys (e.g. cloud adapters)
__all__ = ["GenericSSH", "key_body"]

File: internal/core/providers/ssh_generic.go
package providers

import (
	"context"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	"github.com/simtabi/ssh-manager/internal/core/authkeys"
)

// removeScript is the remote read-modify-write of authorized_keys, run as one ssh
// command under flock. Exit codes: 0 removed; 2 key not present; 3 lockout guard;
// 4 lock/setup failure. {body} is a single-quote-escaped base64 body. Verbatim
// from ssh_generic._REMOVE_SCRIPT.
const removeScript = `set -eu
AK="$HOME/.ssh/authorized_keys"
mkdir -p "$HOME/.ssh"; chmod 700 "$HOME/.ssh"
[ -f "$AK" ] || : > "$AK"
exec 9>"$HOME/.ssh/.ssh-manager.lock" || exit 4
if command -v flock >/dev/null 2>&1; then flock 9 || exit 4; fi
BODY='{body}'
grep -qF -- "$BODY" "$AK" || exit 2
TMP="$(mktemp "$AK.ssh-manager.XXXXXX")" || exit 4
grep -vF -- "$BODY" "$AK" > "$TMP" || true
if ! grep -Eq '^[[:space:]]*[^[:space:]#]' "$TMP"; then rm -f "$TMP"; exit 3; fi
cp -p "$AK" "$AK.ssh-manager.bak.$(date +%Y%m%d-%H%M%S)" 2>/dev/null || true
chmod 600 "$TMP"
mv "$TMP" "$AK"
`

// GenericSSH is the universal fallback: ssh-copy-id / authorized_keys on any
// reachable server. Ports providers.ssh_generic.GenericSSH.
type GenericSSH struct{ spec Spec }

func (g GenericSSH) Name() string {
	if g.spec.Name != "" {
		return g.spec.Name
	}
	return "generic-ssh"
}

func (g GenericSSH) Category() string {
	if g.spec.Category != "" {
		return g.spec.Category
	}
	return "server"
}

func (g GenericSSH) ManageURL(Target) string  { return g.spec.ResolvedKeysURL() }
func (GenericSSH) Rename(Target, string) bool { return false }

func (g GenericSSH) knownHostsOpts(t Target) []string {
	if t.KnownHosts != "" {
		return []string{"-o", "UserKnownHostsFile=" + t.KnownHosts}
	}
	return nil
}

func (g GenericSSH) sshBase(t Target) []string {
	base := append([]string{"ssh", "-o", "StrictHostKeyChecking=accept-new", "-o", "ConnectTimeout=10"}, g.knownHostsOpts(t)...)
	if t.Port != 22 {
		base = append(base, "-p", strconv.Itoa(t.Port))
	}
	return base
}

func (g GenericSSH) Deploy(t Target) DeployOutcome {
	if _, err := exec.LookPath("ssh-copy-id"); err != nil {
		return DeployOutcome{Method: "ssh-copy-id", Detail: "ssh-copy-id not found", Error: true}
	}
	args := append([]string{"-o", "ConnectTimeout=10"}, g.knownHostsOpts(t)...)
	args = append(args, "-i", t.PubkeyPath)
	if t.Port != 22 {
		args = append(args, "-p", strconv.Itoa(t.Port))
	}
	args = append(args, t.SSHDest())
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "ssh-copy-id", args...)
	cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
	if err := cmd.Run(); err != nil {
		return DeployOutcome{Method: "ssh-copy-id", Detail: "ssh-copy-id failed: " + err.Error(), Error: true}
	}
	return DeployOutcome{Method: "ssh-copy-id", Verified: true, Detail: "authorized_keys updated"}
}

func (g GenericSSH) Verify(t Target) bool {
	if t.IdentityPath == "" {
		return false
	}
	args := append([]string{"ssh", "-i", t.IdentityPath, "-o", "BatchMode=yes",
		"-o", "IdentitiesOnly=yes", "-o", "StrictHostKeyChecking=accept-new"}, g.knownHostsOpts(t)...)
	if t.Port != 22 {
		args = append(args, "-p", strconv.Itoa(t.Port))
	}
	args = append(args, t.SSHDest(), "true")
	return runOK(20*time.Second, args...)
}

func (g GenericSSH) ListDeployed(t Target) []string {
	out := g.readAuthorizedKeys(t)
	return authkeys.KeyLines(out)
}

func (g GenericSSH) Remove(t Target) bool {
	body := authkeys.KeyBody(t.PubkeyText)
	if body == "" {
		return false
	}
	script := strings.ReplaceAll(removeScript, "{body}", strings.ReplaceAll(body, "'", "'\\''"))
	args := append(g.sshBase(t), t.SSHDest(), script)
	return runOK(30*time.Second, args...)
}

func (g GenericSSH) readAuthorizedKeys(t Target) string {
	args := append(g.sshBase(t), t.SSHDest(), "cat ~/.ssh/authorized_keys 2>/dev/null || true")
	out, err := runOutput(20*time.Second, args...)
	if err != nil {
		return ""
	}
	return out
}

func runOK(timeout time.Duration, argv ...string) bool {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	return exec.CommandContext(ctx, argv[0], argv[1:]...).Run() == nil
}

func runOutput(timeout time.Duration, argv ...string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, argv[0], argv[1:]...).Output()
	return string(out), err
}
